manta-pay/src/crypto/poseidon/mod.rs

```diff
@@ -0,0 +1,447 @@
+    #[inline]
+    pub fn mds_matrix_multiply(&self, state: &mut State<S, COM>, compiler: &mut COM) {
+        let mut next = Vec::with_capacity(S::WIDTH);
+        for i in 0..S::WIDTH {
+            // NOTE: clippy false-positive: Without `collect`, the two closures in `map` and
+            //       `reduce` will have simultaneous `&mut` access to `compiler`. Adding `collect`
+            //       allows `map` to be done before `reduce`.
+            #[allow(clippy::needless_collect)]
+            let linear_combination = state
+                .iter()
+                .enumerate()
+                .map(|(j, elem)| S::mul_const(elem, &self.mds_matrix[S::WIDTH * i + j], compiler))
+                .collect::<Vec<_>>();
+            next.push(
+                linear_combination
+                    .into_iter()
+                    .reduce(|acc, next| S::add(&acc, &next, compiler))
+                    .unwrap(),
+            );
+        }
+        mem::swap(&mut next.into_boxed_slice(), &mut state.0);
+    }
+
+    /// Computes a full round at the given `round` index on the internal permutation `state`.
+    #[inline]
+    pub fn full_round(&self, round: usize, state: &mut State<S, COM>, compiler: &mut COM) {
+        let keys = self.additive_keys(round);
+        for (i, elem) in state.iter_mut().enumerate() {
+            S::add_const_assign(elem, &keys[i], compiler);
+            S::apply_sbox(elem, compiler);
+        }
+        self.mds_matrix_multiply(state, compiler);
+    }
+
+    /// Computes a partial round at the given `round` index on the internal permutation `state`.
+    #[inline]
+    pub fn partial_round(&self, round: usize, state: &mut State<S, COM>, compiler: &mut COM) {
+        let keys = self.additive_keys(round);
+        for (i, elem) in state.iter_mut().enumerate() {
+            S::add_const_assign(elem, &keys[i], compiler);
+        }
+        S::apply_sbox(&mut state.0[0], compiler);
+        self.mds_matrix_multiply(state, compiler);
+    }
+
+    /// Computes the full permutation without the first round.
+    #[inline]
+    fn permute_without_first_round(&self, state: &mut State<S, COM>, compiler: &mut COM) {
+        for round in 1..Self::HALF_FULL_ROUNDS {
+            self.full_round(round, state, compiler);
+        }
+        for round in Self::HALF_FULL_ROUNDS..(Self::HALF_FULL_ROUNDS + S::PARTIAL_ROUNDS) {
+            self.partial_round(round, state, compiler);
+        }
+        for round in
+            (Self::HALF_FULL_ROUNDS + S::PARTIAL_ROUNDS)..(S::FULL_ROUNDS + S::PARTIAL_ROUNDS)
+        {
+            self.full_round(round, state, compiler);
+        }
+    }
+
+    /// Computes the first round borrowing the `input` and `domain_tag` returning the [`State`]
+    /// after the first round. This method does not check that `N + 1 = S::WIDTH`.
+    #[inline]
+    fn first_round_with_domain_tag_unchecked<const N: usize>(
+        &self,
+        domain_tag: &S::Field,
```

Comment:
`domain_tag` could be `S::ParameterField`.

Reply:
Not really? Domain tag is part of initial state. `ParameterField` should only be used in round constants and mds.

Reply:
`domain_tag` is a part of the specification that is public. It is also fixed for a specific kind of Poseidon hash. So it could be `S::ParameterField`.

Reply:
Then `iter::once(domain_tag).chain(input)` will not compile (Incompatible type)

Reply:
We can write another function that can pass `domain_tag` as a `ParameterField` to have both options. It could be that `domain_tag` is `Field` but it is compared equal to some `ParameterField` later (there are many ways to fix the domain tag in the circuit other than just having it as a constant). But let's try to make it a constant at this level.